pfcwd/main.py

```diff
@@ -32,272 +53,438 @@
+
+        if not enable or enable.lower() != "enable":
+            return
+
+        port_num = len(self.config_db.get_table('PORT').keys())
+
+        # Paramter values positively correlate to the number of ports.
+        multiply = max(1, (port_num-1)/DEFAULT_PORT_NUM+1)
+        pfcwd_info = {
+            'detection_time': DEFAULT_DETECTION_TIME * multiply,
+            'restoration_time': DEFAULT_RESTORATION_TIME * multiply,
+            'action': DEFAULT_ACTION
+        }
+
+        for port in active_ports:
+            self.config_db.set_entry(
+                CONFIG_DB_PFC_WD_TABLE_NAME, port, pfcwd_info
+            )
+
+        pfcwd_info = {}
+        pfcwd_info['POLL_INTERVAL'] = DEFAULT_POLL_INTERVAL * multiply
+        self.config_db.mod_entry(
+            CONFIG_DB_PFC_WD_TABLE_NAME, "GLOBAL", pfcwd_info
+        )
+
+    @multi_asic_util.run_on_multi_asic
+    def counter_poll(self, counter_poll):
+        if os.geteuid() != 0:
+            exit("Root privileges are required for this operation")
+        pfcwd_info = {}
+        pfcwd_info['FLEX_COUNTER_STATUS'] = counter_poll
+        self.config_db.mod_entry("FLEX_COUNTER_TABLE", "PFCWD", pfcwd_info)
+
+    @multi_asic_util.run_on_multi_asic
+    def big_red_switch(self, big_red_switch):
+        if os.geteuid() != 0:
+            exit("Root privileges are required for this operation")
+        pfcwd_info = {}
+        if big_red_switch is not None:
+            pfcwd_info['BIG_RED_SWITCH'] = big_red_switch
+        self.config_db.mod_entry(
+            CONFIG_DB_PFC_WD_TABLE_NAME, "GLOBAL",
+            pfcwd_info
+        )
+
 
 # Show stats
-@show.command()
-@click.argument('ports', nargs = -1)
-def config(ports):
-    """ Show PFC Watchdog configuration """
-    configdb = swsssdk.ConfigDBConnector()
-    configdb.connect()
-    countersdb = swsssdk.SonicV2Connector(host='127.0.0.1')
-    countersdb.connect(countersdb.COUNTERS_DB)
-    table = []
-
-    all_ports = get_all_ports(countersdb)
-
-    if len(ports) == 0:
-        ports = all_ports
-
-    for port in ports:
-        config_list = []
-        config_entry = configdb.get_entry(CONFIG_DB_PFC_WD_TABLE_NAME, port)
-        if config_entry is None or config_entry == {}:
-            continue
-        for config in CONFIG_DESCRIPTION:
-            line = config_entry.get(config[1], config[2])
-            config_list.append(line)
-        table.append([port] + config_list)
-    poll_interval = configdb.get_entry( CONFIG_DB_PFC_WD_TABLE_NAME, 'GLOBAL').get('POLL_INTERVAL')
-    if poll_interval is not None:
-        click.echo("Changed polling interval to " + poll_interval + "ms")
-
-    big_red_switch = configdb.get_entry( CONFIG_DB_PFC_WD_TABLE_NAME, 'GLOBAL').get('BIG_RED_SWITCH')
-    if big_red_switch is not None:
-        click.echo("BIG_RED_SWITCH status is " + big_red_switch)
-
-    click.echo(tabulate(table, CONFIG_HEADER, stralign='right', numalign='right', tablefmt='simple'))
+class Show(object):
```

Comment:
What is the benefit in encapsulating these groups inside classes?

Reply:
Main benefit is adding sub commands becomes easier and are logically grouped. Besides, multi ASIC decorator requires a class instance to operate upon. I also thought of adding a base class like 'SonicCli' but then decided against it as the only user is 'multi ASIC' decorator which does not apply to all the CLIs.